Broken TypeScript generated for multipart/form-data request body with oneOf/anyOf

## Description

When an OpenAPI 3.1 spec uses `oneOf` (or `anyOf`) inside a `multipart/form-data` request body, Orval generates broken TypeScript code that fails type-checking. There are two distinct, independent bugs in `packages/core/src/getters/res-req-types.ts`:

- **Bug 1**: Variable name mismatch in `getResReqTypes`. The FormData code references a variable that doesn't exist. TypeScript compilation fails. Triggers as soon as a `oneOf` is present, even with a single schema inside.
- **Bug 2**: Duplicated `formData.append()` calls in `getSchemaFormDataAndUrlEncoded`. All `oneOf` branches run unconditionally, doubling shared fields at runtime. Triggers when the `oneOf` contains 2 or more schemas.

This is related to [#2747](https://github.com/orval-labs/orval/issues/2747) (fixed in 8.0.1), but that fix only addressed `multipart/form-data` **without** `oneOf`/`anyOf`. The `oneOf` code path was not updated and is still broken as of 8.8.0.

JSON `application/json` endpoints with `oneOf` are **not** affected because they use `JSON.stringify()` and skip the FormData generation entirely.

**Note:** `oneOf` with a single entry is a common pattern emitted by NestJS decorators (`@nestjs/swagger`, `zod-to-openapi`) for versioned request schemas. The spec is valid per OpenAPI 3.1.

## Minimal reproduction

### OpenAPI spec

```yaml
openapi: 3.1.0
info:
  title: Test API
  version: 1.0.0
paths:
  /upload:
    post:
      operationId: FileController_upload
      requestBody:
        required: true
        content:
          multipart/form-data:
            schema:
              oneOf:
                - $ref: "#/components/schemas/UploadFileDto"
      responses:
        "201":
          description: Created
  /upload-versioned:
    post:
      operationId: FileController_uploadVersioned
      requestBody:
        required: true
        content:
          multipart/form-data:
            schema:
              oneOf:
                - $ref: "#/components/schemas/UploadFileDtoV1"
                - $ref: "#/components/schemas/UploadFileDtoV2"
      responses:
        "201":
          description: Created
components:
  schemas:
    UploadFileDto:
      type: object
      properties:
        file:
          type: string
          format: binary
        label:
          type: string
        category:
          type: string
      required: [file, label, category]
    UploadFileDtoV1:
      type: object
      properties:
        file:
          type: string
          format: binary
        label:
          type: string
        category:
          type: string
      required: [file, label, category]
    UploadFileDtoV2:
      type: object
      properties:
        file:
          type: string
          format: binary
        label:
          type: string
        category:
          type: string
        tags:
          type: string
      required: [file, label, category, tags]
```

### Orval config

```ts
import { defineConfig } from 'orval';

export default defineConfig({
  api: {
    output: {
      target: './generated.ts',
      client: 'fetch',
    },
    input: {
      target: './openapi.yaml',
    },
  },
});
```

### Steps to reproduce

1. Save the spec and config above
2. Run `npx orval --config orval.config.ts`
3. Run `npx tsc --noEmit --strict --lib es2020,dom generated.ts`
4. Observe compilation errors on both endpoints

## Bug 1: Variable name mismatch in `getResReqTypes` (even with a single schema in `oneOf`)

### Root cause

Within `getResReqTypes`, there is a mismatch in how the variable name is resolved depending on whether the schema is a direct `$ref` or a `oneOf` wrapper:

- The **type resolution** path (via `getResReqContentTypes` / `resolveObject`) correctly walks through the `oneOf` and resolves the `$ref` inside. This produces the correct function parameter name (e.g. `uploadFileDto`).
- The **`effectivePropName`** path (used for the FormData variable) only checks `isReference(mediaType.schema)`. When the schema is `{ oneOf: [{ $ref: "..." }] }`, this check returns `false` because the top-level object is not a `$ref`. So `effectivePropName` falls back to the default name derived from the operation (e.g. `FileControllerUploadBody`).

Result: the function parameter is named `uploadFileDto` but the FormData code references `fileControllerUploadBody`, which was never declared.

### TypeScript compiler error

```
generated.ts(48,48): error TS2552: Cannot find name 'fileControllerUploadBody'. Did you mean 'fileControllerUpload'?
generated.ts(93,59): error TS2552: Cannot find name 'fileControllerUploadVersionedBody'. Did you mean 'fileControllerUploadVersioned'?
generated.ts(98,59): error TS2552: Cannot find name 'fileControllerUploadVersionedBody'. Did you mean 'fileControllerUploadVersioned'?
```

### Expected generated code

```typescript
export const fileControllerUpload = async (
  uploadFileDto: UploadFileDto
) => {
  const formData = new FormData();
  formData.append('file', uploadFileDto.file);
  //                       ^^^^^^^^^^^^^ matches the parameter name
```

### Actual generated code

```typescript
export const fileControllerUpload = async (
  uploadFileDto: UploadFileDto
) => {
  const formData = new FormData();
  formData.append('file', fileControllerUploadBody.file);
  //                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^ never declared
```

## Bug 2: Duplicated `formData.append()` calls (when `oneOf` contains 2+ schemas)

This is a separate bug from Bug 1. It manifests when there are 2+ schemas in the `oneOf`:

```yaml
oneOf:
  - $ref: "#/components/schemas/UploadFileDtoV1"
  - $ref: "#/components/schemas/UploadFileDtoV2"
```

### Root cause

In `getSchemaFormDataAndUrlEncoded`, when the schema has `combinedSchemas` (oneOf/anyOf), the code maps over each sub-schema and generates a `formData.append()` block for every field, then joins all blocks with `\n`. All blocks run unconditionally. There is no `if/else` or discriminator logic between them.

The TypeScript `as` casts (e.g. `param as UploadFileDtoV1`) are compile-time only and disappear at runtime, so both variables reference the same object. Every shared field gets appended to the FormData twice.

### Expected behavior

Each field should only be appended once at runtime, regardless of which `oneOf` variant is passed.

### Actual generated code (with 2 schemas in oneOf)

```typescript
const formData = new FormData();

// Block for V1 (always runs)
const uploadFileDtoV1 = (param as UploadFileDtoV1);
formData.append('file', uploadFileDtoV1.file);
formData.append('label', uploadFileDtoV1.label);
formData.append('category', uploadFileDtoV1.category);

// Block for V2 (also always runs, no condition!)
const uploadFileDtoV2 = (param as UploadFileDtoV2);
formData.append('file', uploadFileDtoV2.file);          // DUPLICATE
formData.append('label', uploadFileDtoV2.label);        // DUPLICATE
formData.append('category', uploadFileDtoV2.category);  // DUPLICATE
formData.append('tags', uploadFileDtoV2.tags);
```

## Workaround

We're using a `pnpm patchedDependencies` patch on `@orval/core@8.4.2` that fixes both bugs.

## Suggested fix direction

Two changes needed in `packages/core/src/getters/res-req-types.ts`:

### Bug 1: resolve `effectivePropName` through `oneOf`/`anyOf`

In `getResReqTypes`, after the existing `if (isReference(mediaType.schema))` block, add a branch that unwraps the `oneOf`/`anyOf`:

```typescript
// Existing code handles direct $ref:
if (isReference(mediaType.schema)) {
  // ... resolves effectivePropName correctly
}
// Add: resolve through oneOf/anyOf wrapper
else if (mediaType.schema) {
  const combinedRef = mediaType.schema.oneOf ?? mediaType.schema.anyOf;
  if (combinedRef) {
    const names = combinedRef
      .filter((ref) => isReference(ref))
      .map((ref) => resolveRef(ref, context).imports[0]?.name)
      .filter(Boolean);
    if (names.length) {
      effectivePropName = names.join('');
    }
  }
}
```

This resolves the `$ref`(s) inside the `oneOf` and uses their names. With a single schema, it gives the correct DTO name. With multiple schemas, it concatenates the names (same behavior as the existing JSON code path).

### Bug 2: generate a single FormData block instead of one per `oneOf` variant

In `getSchemaFormDataAndUrlEncoded`, for `oneOf`/`anyOf` (`shouldCast === true`), replace the `.map().join('\n')` pattern (which generates all variant blocks unconditionally) with a single `Object.entries()` loop that iterates over the object's actual properties at runtime:

```typescript
if (shouldCast) {
  form += `Object.entries(${propName}).forEach(([key, value]) => {\n`;
  form += `  if (value !== undefined) {\n`;
  form += `    if (value instanceof File || value instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(value))) {\n`;
  form += `      ${variableName}.append(key, value);\n`;
  form += `    } else if (Array.isArray(value)) {\n`;
  form += `      value.forEach(v => ${variableName}.append(key, typeof v === 'object' ? JSON.stringify(v) : String(v)));\n`;
  form += `    } else if (typeof value === 'object' && value !== null) {\n`;
  form += `      ${variableName}.append(key, JSON.stringify(value));\n`;
  form += `    } else {\n`;
  form += `      ${variableName}.append(key, String(value));\n`;
  form += `    }\n`;
  form += `  }\n`;
  form += `});\n`;

  // Preserve imports from all sub-schemas
  combinedSchemas.forEach((subSchema) => {
    const { imports } = resolveRef(subSchema, context);
    if (imports[0]) additionalImports.push(imports[0]);
  });
} else {
  // allOf case: preserve original behavior (unchanged)
}
```

The type checks (File/Blob/Buffer, Array, Object, String) match the cases handled by `resolveSchemaPropertiesToFormData`, the function Orval already uses for multipart endpoints **without** `oneOf`.

**Known limitation**: the `Object.entries()` loop iterates dynamically at runtime, so it doesn't know each field's schema. Orval's existing static path (`resolveSchemaPropertiesToFormData`) generates one `formData.append()` per field and can skip `readOnly` fields, apply per-field `encoding` overrides, or serialize arrays differently. Our dynamic loop can't do that because it doesn't have that per-field information at runtime.

This works for our use case (no `readOnly`, no `encoding` overrides), but if you'd prefer a fix that preserves those features, we're open to a different approach.

## Environment

- **Orval version**: tested on 8.4.2 and reproduced on 8.8.0 (same broken output, same TS compilation errors)
- **Client**: `fetch` (the bug is in `getResReqTypes` / `getSchemaFormDataAndUrlEncoded` which are shared across clients, so `axios`, `react-query`, etc. are likely affected too, but we only tested `fetch`)
- **OpenAPI version**: 3.1.0

## Related issues

- [#2747](https://github.com/orval-labs/orval/issues/2747): Fixed `multipart/form-data` handling in 8.0.1, but only for schemas **without** `oneOf`/`anyOf`. The `oneOf` code path was not updated.

---

Happy to open a PR if this direction works for you. We have a working patch with this approach and can adapt it to the TS source with tests.

@melloware

--- a/packages/core/src/getters/res-req-types.test.ts
+++ b/packages/core/src/getters/res-req-types.test.ts
@@ -725,6 +725,79 @@ bodyRequestBody.photos.forEach(value => formData.append(\`photos\`, value));
       );
     });
 
+    it('oneOf alongside direct properties: loop skips direct keys to avoid duplicate appends', () => {
+      const ctx: ContextSpec = {
+        ...context,
+        spec: {
+          components: {
+            schemas: {
+              VariantA: {
+                type: 'object',
+                properties: {
+                  kind: { type: 'string', enum: ['a'] },
+                  extraA: { type: 'string' },
+                },
+                required: ['kind'],
+              },
+              VariantB: {
+                type: 'object',
+                properties: {
+                  kind: { type: 'string', enum: ['b'] },
+                  extraB: { type: 'string' },
+                },
+                required: ['kind'],
+              },
+            },
+          },
+        },
+      };
+
+      const reqBody: [string, OpenApiRequestBodyObject][] = [
+        [
+          'requestBody',
+          {
+            content: {
+              'multipart/form-data': {
+                schema: {
+                  type: 'object',
+                  oneOf: [
+                    { $ref: '#/components/schemas/VariantA' },
+                    { $ref: '#/components/schemas/VariantB' },
+                  ],
+                  properties: {
+                    name: { type: 'string' },
+                    owner: { type: 'string' },
+                  },
+                  required: ['name'],
+                },
+              },
+            },
+            required: true,
+          },
+        ],
+      ];
+
+      const result = getResReqTypes(reqBody, 'Upload', ctx)[0];
+      const formData = result.formData;
+      if (!formData || !isString(formData)) {
+        throw new Error('Expected formData to be a defined string');
+      }
+
+      // The Object.entries loop skips keys that the direct-properties branch
+      // already appends (`name`, `owner`), so those fields are not appended
+      // twice at runtime.
+      expect(formData).toContain('Object.entries(');
+      expect(formData).toContain('["name", "owner"].includes(key)');
+      const nameAppendCount = (
+        formData.match(/formData\.append\(`name`/g) ?? []
+      ).length;
+      const ownerAppendCount = (
+        formData.match(/formData\.append\(`owner`/g) ?? []
+      ).length;
+      expect(nameAppendCount).toBe(1);
+      expect(ownerAppendCount).toBe(1);
+    });
+
     it('oneOf with an optional body: FormData guards against undefined', () => {
       const ctx: ContextSpec = {
         ...context,

--- a/packages/core/src/getters/res-req-types.test.ts
+++ b/packages/core/src/getters/res-req-types.test.ts
@@ -722,6 +722,112 @@ bodyRequestBody.photos.forEach(value => formData.append(\`photos\`, value));
         'formData.append(`logo`, uploadRequestBody.logo)',
       );
     });
+
+    it('oneOf with an optional body: FormData guards against undefined', () => {
+      const ctx: ContextSpec = {
+        ...context,
+        spec: {
+          components: {
+            schemas: {
+              OptionalBodyDto: {
+                type: 'object',
+                properties: {
+                  file: { type: 'string', format: 'binary' },
+                },
+              },
+            },
+          },
+        },
+      };
+
+      const reqBody: [string, OpenApiRequestBodyObject][] = [
+        [
+          'requestBody',
+          {
+            content: {
+              'multipart/form-data': {
+                schema: {
+                  oneOf: [{ $ref: '#/components/schemas/OptionalBodyDto' }],
+                },
+              },
+            },
+            required: false,
+          },
+        ],
+      ];
+
+      const result = getResReqTypes(reqBody, 'Upload', ctx)[0];
+      const formData = result.formData;
+      if (!formData || !isString(formData)) {
+        throw new Error('Expected formData to be a defined string');
+      }
+
+      // The runtime loop must defend against an undefined body so callers can
+      // safely pass nothing when the body is optional.
+      expect(formData).toMatch(/Object\.entries\([^)]*\?\?\s*\{\}\)/);
+    });
+
+    it('oneOf with an array of binary files: FormData appends Blob items directly', () => {
+      const ctx: ContextSpec = {
+        ...context,
+        spec: {
+          components: {
+            schemas: {
+              MultiUploadV1: {
+                type: 'object',
+                properties: {
+                  files: {
+                    type: 'array',
+                    items: { type: 'string', format: 'binary' },
+                  },
+                },
+              },
+              MultiUploadV2: {
+                type: 'object',
+                properties: {
+                  files: {
+                    type: 'array',
+                    items: { type: 'string', format: 'binary' },
+                  },
+                  tag: { type: 'string' },
+                },
+              },
+            },
+          },
+        },
+      };
+
+      const reqBody: [string, OpenApiRequestBodyObject][] = [
+        [
+          'requestBody',
+          {
+            content: {
+              'multipart/form-data': {
+                schema: {
+                  oneOf: [
+                    { $ref: '#/components/schemas/MultiUploadV1' },
+                    { $ref: '#/components/schemas/MultiUploadV2' },
+                  ],
+                },
+              },
+            },
+            required: true,
+          },
+        ],
+      ];
+
+      const result = getResReqTypes(reqBody, 'Upload', ctx)[0];
+      const formData = result.formData;
+      if (!formData || !isString(formData)) {
+        throw new Error('Expected formData to be a defined string');
+      }
+
+      // Array elements that are files must be appended directly. Without the
+      // per-element Blob/File/Buffer check, each binary item would have been
+      // serialized to "{}" via JSON.stringify.
+      expect(formData).toContain('value.forEach');
+      expect(formData).toContain('v instanceof Blob');
+    });
   });
 });
 

--- a/packages/core/src/getters/res-req-types.test.ts
+++ b/packages/core/src/getters/res-req-types.test.ts
@@ -657,9 +657,11 @@ bodyRequestBody.photos.forEach(value => formData.append(\`photos\`, value));
       // Shared fields used to be appended once per variant because TypeScript
       // casts are erased at runtime. The runtime loop appends each key once.
       expect(formData).toContain('Object.entries(');
-      const appendFileCount = (formData.match(/`file`/g) ?? []).length;
-      expect(appendFileCount).toBe(0);
       expect(formData).not.toMatch(/as UploadDtoV[12]/);
+      // Guard against a future refactor reintroducing the per-variant
+      // appends keyed off the shared field.
+      expect(formData).not.toContain('uploadDtoV1.file');
+      expect(formData).not.toContain('uploadDtoV2.file');
 
       // Variant types are still imported so they remain referenceable.
       const importNames = result.imports.map((i) => i.name);

--- a/packages/core/src/getters/res-req-types.test.ts
+++ b/packages/core/src/getters/res-req-types.test.ts
@@ -547,6 +547,181 @@ bodyRequestBody.photos.forEach(value => formData.append(\`photos\`, value));
       // allOf with $ref: intersection type (not union)
       expect(schema?.model).toContain('ClientUpdateDto & {');
     });
+
+    // Regression tests for #3242: multipart/form-data with oneOf/anyOf
+    // at the root of a request body (common with @nestjs/swagger or
+    // zod-to-openapi for versioned request schemas).
+    it('oneOf with a single $ref: FormData variable derives from the DTO name', () => {
+      const ctx: ContextSpec = {
+        ...context,
+        spec: {
+          components: {
+            schemas: {
+              ClientUpdateDto: {
+                type: 'object',
+                properties: {
+                  name: { type: 'string' },
+                  logo: { type: 'string', format: 'binary' },
+                },
+                required: ['name', 'logo'],
+              },
+            },
+          },
+        },
+      };
+
+      const reqBody: [string, OpenApiRequestBodyObject][] = [
+        [
+          'requestBody',
+          {
+            content: {
+              'multipart/form-data': {
+                schema: {
+                  oneOf: [{ $ref: '#/components/schemas/ClientUpdateDto' }],
+                },
+              },
+            },
+            required: true,
+          },
+        ],
+      ];
+
+      const result = getResReqTypes(reqBody, 'Upload', ctx)[0];
+      const formData = result.formData;
+      if (!formData || !isString(formData)) {
+        throw new Error('Expected formData to be a defined string');
+      }
+
+      // Without the fix, effectivePropName fell back to the controller-derived
+      // name and the generated code referenced a variable that was never
+      // declared (e.g. uploadRequestBody).
+      expect(formData).toContain('clientUpdateDto');
+      expect(formData).not.toContain('uploadRequestBody');
+    });
+
+    it('oneOf with 2 $refs: FormData uses a runtime Object.entries loop', () => {
+      const ctx: ContextSpec = {
+        ...context,
+        spec: {
+          components: {
+            schemas: {
+              UploadDtoV1: {
+                type: 'object',
+                properties: {
+                  file: { type: 'string', format: 'binary' },
+                  metadata: { type: 'string' },
+                },
+                required: ['file'],
+              },
+              UploadDtoV2: {
+                type: 'object',
+                properties: {
+                  file: { type: 'string', format: 'binary' },
+                  metadata: {
+                    type: 'object',
+                    properties: { name: { type: 'string' } },
+                  },
+                },
+                required: ['file'],
+              },
+            },
+          },
+        },
+      };
+
+      const reqBody: [string, OpenApiRequestBodyObject][] = [
+        [
+          'requestBody',
+          {
+            content: {
+              'multipart/form-data': {
+                schema: {
+                  oneOf: [
+                    { $ref: '#/components/schemas/UploadDtoV1' },
+                    { $ref: '#/components/schemas/UploadDtoV2' },
+                  ],
+                },
+              },
+            },
+            required: true,
+          },
+        ],
+      ];
+
+      const result = getResReqTypes(reqBody, 'Upload', ctx)[0];
+      const formData = result.formData;
+      if (!formData || !isString(formData)) {
+        throw new Error('Expected formData to be a defined string');
+      }
+
+      // Shared fields used to be appended once per variant because TypeScript
+      // casts are erased at runtime. The runtime loop appends each key once.
+      expect(formData).toContain('Object.entries(');
+      const appendFileCount = (formData.match(/`file`/g) ?? []).length;
+      expect(appendFileCount).toBe(0);
+      expect(formData).not.toMatch(/as UploadDtoV[12]/);
+
+      // Variant types are still imported so they remain referenceable.
+      const importNames = result.imports.map((i) => i.name);
+      expect(importNames).toContain('UploadDtoV1');
+      expect(importNames).toContain('UploadDtoV2');
+    });
+
+    it('allOf: FormData still emits per-field appends (no regression)', () => {
+      const ctx: ContextSpec = {
+        ...context,
+        spec: {
+          components: {
+            schemas: {
+              BaseDto: {
+                type: 'object',
+                properties: { name: { type: 'string' } },
+                required: ['name'],
+              },
+            },
+          },
+        },
+      };
+
+      const reqBody: [string, OpenApiRequestBodyObject][] = [
+        [
+          'requestBody',
+          {
+            content: {
+              'multipart/form-data': {
+                schema: {
+                  allOf: [
+                    { $ref: '#/components/schemas/BaseDto' },
+                    {
+                      type: 'object',
+                      properties: {
+                        logo: { type: 'string', format: 'binary' },
+                      },
+                      required: ['logo'],
+                    },
+                  ],
+                },
+              },
+            },
+            required: true,
+          },
+        ],
+      ];
+
+      const result = getResReqTypes(reqBody, 'Upload', ctx)[0];
+      const formData = result.formData;
+      if (!formData || !isString(formData)) {
+        throw new Error('Expected formData to be a defined string');
+      }
+
+      expect(formData).not.toContain('Object.entries(');
+      expect(formData).toContain(
+        'formData.append(`name`, uploadRequestBody.name)',
+      );
+      expect(formData).toContain(
+        'formData.append(`logo`, uploadRequestBody.logo)',
+      );
+    });
   });
 });
 

--- a/packages/core/src/getters/res-req-types.ts
+++ b/packages/core/src/getters/res-req-types.ts
@@ -576,14 +576,31 @@ function getSchemaFormDataAndUrlEncoded({
       const shouldCast = !!getSchemaOneOf(schema) || !!getSchemaAnyOf(schema);
 
       if (shouldCast) {
+        // If the outer schema also has direct properties, those are handled
+        // below by the dedicated properties branch. Skip them here to avoid
+        // appending the same key twice.
+        const directProperties = getSchemaProperties(schema);
+        const directKeys = directProperties
+          ? Object.keys(directProperties)
+          : [];
+        const skipLine =
+          directKeys.length > 0
+            ? `  if ([${directKeys.map((k) => JSON.stringify(k)).join(', ')}].includes(key)) return;\n`
+            : '';
+
         form += `Object.entries(${propName} ?? {}).forEach(([key, value]) => {\n`;
+        form += skipLine;
         form += `  if (value !== undefined && value !== null) {\n`;
-        form += `    if ((typeof File !== 'undefined' && value instanceof File) || value instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(value))) {\n`;
+        form += `    if ((typeof File !== 'undefined' && value instanceof File) || value instanceof Blob) {\n`;
         form += `      ${variableName}.append(key, value);\n`;
+        form += `    } else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(value)) {\n`;
+        form += `      ${variableName}.append(key, new Blob([value as unknown as BlobPart]));\n`;
         form += `    } else if (Array.isArray(value)) {\n`;
         form += `      value.forEach(v => {\n`;
-        form += `        if ((typeof File !== 'undefined' && v instanceof File) || v instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(v))) {\n`;
+        form += `        if ((typeof File !== 'undefined' && v instanceof File) || v instanceof Blob) {\n`;
         form += `          ${variableName}.append(key, v);\n`;
+        form += `        } else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(v)) {\n`;
+        form += `          ${variableName}.append(key, new Blob([v as unknown as BlobPart]));\n`;
         form += `        } else {\n`;
         form += `          ${variableName}.append(key, typeof v === 'object' ? JSON.stringify(v) : String(v));\n`;
         form += `        }\n`;

--- a/packages/core/src/getters/res-req-types.ts
+++ b/packages/core/src/getters/res-req-types.ts
@@ -257,7 +257,6 @@ export function getResReqTypes(
                 effectivePropName = imports[0].name;
               }
             } else if (mediaType.schema) {
-              // When schema is a oneOf/anyOf of $refs, concat schema names for consistent param naming
               const combinedRefs =
                 getSchemaOneOf(mediaType.schema) ??
                 getSchemaAnyOf(mediaType.schema);
@@ -265,9 +264,10 @@ export function getResReqTypes(
                 const names: string[] = [];
                 for (const ref of combinedRefs) {
                   if (!isReference(ref)) continue;
-                  const name = resolveSchemaRef(ref, context).imports[0]?.name;
-                  if (name) {
-                    names.push(name);
+                  const refName = resolveSchemaRef(ref, context).imports[0]
+                    ?.name;
+                  if (refName) {
+                    names.push(refName);
                   }
                 }
                 if (names.length > 0) {
@@ -561,7 +561,6 @@ function getSchemaFormDataAndUrlEncoded({
   const propName = camel(
     !isRef && isReference(schemaObject) ? imports[0].name : name,
   );
-  const additionalImports: GeneratorImport[] = [];
 
   const variableName = isUrlEncoded ? 'formUrlEncoded' : 'formData';
   let form = isUrlEncoded
@@ -577,28 +576,25 @@ function getSchemaFormDataAndUrlEncoded({
       const shouldCast = !!getSchemaOneOf(schema) || !!getSchemaAnyOf(schema);
 
       if (shouldCast) {
-        // oneOf/anyOf: iterate over actual properties at runtime to avoid duplicating
-        // fields shared across variants (TypeScript casts are erased at compile-time).
-        form += `Object.entries(${propName}).forEach(([key, value]) => {\n`;
-        form += `  if (value !== undefined) {\n`;
-        form += `    if (value instanceof File || value instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(value))) {\n`;
+        form += `Object.entries(${propName} ?? {}).forEach(([key, value]) => {\n`;
+        form += `  if (value !== undefined && value !== null) {\n`;
+        form += `    if ((typeof File !== 'undefined' && value instanceof File) || value instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(value))) {\n`;
         form += `      ${variableName}.append(key, value);\n`;
         form += `    } else if (Array.isArray(value)) {\n`;
-        form += `      value.forEach(v => ${variableName}.append(key, typeof v === 'object' ? JSON.stringify(v) : String(v)));\n`;
-        form += `    } else if (typeof value === 'object' && value !== null) {\n`;
+        form += `      value.forEach(v => {\n`;
+        form += `        if ((typeof File !== 'undefined' && v instanceof File) || v instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(v))) {\n`;
+        form += `          ${variableName}.append(key, v);\n`;
+        form += `        } else {\n`;
+        form += `          ${variableName}.append(key, typeof v === 'object' ? JSON.stringify(v) : String(v));\n`;
+        form += `        }\n`;
+        form += `      });\n`;
+        form += `    } else if (typeof value === 'object') {\n`;
         form += `      ${variableName}.append(key, JSON.stringify(value));\n`;
         form += `    } else {\n`;
         form += `      ${variableName}.append(key, String(value));\n`;
         form += `    }\n`;
         form += `  }\n`;
         form += `});\n`;
-
-        for (const subSchema of combinedSchemas) {
-          const { imports } = resolveSchemaRef(subSchema, context);
-          if (imports[0]) {
-            additionalImports.push(imports[0]);
-          }
-        }
       } else {
         const combinedSchemasFormData = combinedSchemas
           .map((subSchema) => {

--- a/packages/core/src/getters/res-req-types.ts
+++ b/packages/core/src/getters/res-req-types.ts
@@ -594,13 +594,13 @@ function getSchemaFormDataAndUrlEncoded({
         form += `    if ((typeof File !== 'undefined' && value instanceof File) || value instanceof Blob) {\n`;
         form += `      ${variableName}.append(key, value);\n`;
         form += `    } else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(value)) {\n`;
-        form += `      ${variableName}.append(key, new Blob([value as unknown as BlobPart]));\n`;
+        form += `      ${variableName}.append(key, new Blob([Uint8Array.from(value)]));\n`;
         form += `    } else if (Array.isArray(value)) {\n`;
         form += `      value.forEach(v => {\n`;
         form += `        if ((typeof File !== 'undefined' && v instanceof File) || v instanceof Blob) {\n`;
         form += `          ${variableName}.append(key, v);\n`;
         form += `        } else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(v)) {\n`;
-        form += `          ${variableName}.append(key, new Blob([v as unknown as BlobPart]));\n`;
+        form += `          ${variableName}.append(key, new Blob([Uint8Array.from(v)]));\n`;
         form += `        } else {\n`;
         form += `          ${variableName}.append(key, typeof v === 'object' ? JSON.stringify(v) : String(v));\n`;
         form += `        }\n`;

--- a/packages/core/src/getters/res-req-types.ts
+++ b/packages/core/src/getters/res-req-types.ts
@@ -256,6 +256,24 @@ export function getResReqTypes(
               if (imports[0]?.name) {
                 effectivePropName = imports[0].name;
               }
+            } else if (mediaType.schema) {
+              // When schema is a oneOf/anyOf of $refs, concat schema names for consistent param naming
+              const combinedRefs =
+                getSchemaOneOf(mediaType.schema) ??
+                getSchemaAnyOf(mediaType.schema);
+              if (combinedRefs) {
+                const names: string[] = [];
+                for (const ref of combinedRefs) {
+                  if (!isReference(ref)) continue;
+                  const name = resolveSchemaRef(ref, context).imports[0]?.name;
+                  if (name) {
+                    names.push(name);
+                  }
+                }
+                if (names.length > 0) {
+                  effectivePropName = names.join('');
+                }
+              }
             }
 
             const isFormData = formDataContentTypes.has(contentType);
@@ -558,40 +576,50 @@ function getSchemaFormDataAndUrlEncoded({
     if (combinedSchemas) {
       const shouldCast = !!getSchemaOneOf(schema) || !!getSchemaAnyOf(schema);
 
-      const combinedSchemasFormData = combinedSchemas
-        .map((subSchema) => {
-          const { schema: combinedSchema, imports } = resolveSchemaRef(
-            subSchema,
-            context,
-          );
+      if (shouldCast) {
+        // oneOf/anyOf: iterate over actual properties at runtime to avoid duplicating
+        // fields shared across variants (TypeScript casts are erased at compile-time).
+        form += `Object.entries(${propName}).forEach(([key, value]) => {\n`;
+        form += `  if (value !== undefined) {\n`;
+        form += `    if (value instanceof File || value instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(value))) {\n`;
+        form += `      ${variableName}.append(key, value);\n`;
+        form += `    } else if (Array.isArray(value)) {\n`;
+        form += `      value.forEach(v => ${variableName}.append(key, typeof v === 'object' ? JSON.stringify(v) : String(v)));\n`;
+        form += `    } else if (typeof value === 'object' && value !== null) {\n`;
+        form += `      ${variableName}.append(key, JSON.stringify(value));\n`;
+        form += `    } else {\n`;
+        form += `      ${variableName}.append(key, String(value));\n`;
+        form += `    }\n`;
+        form += `  }\n`;
+        form += `});\n`;
 
-          let newPropName = propName;
-          let newPropDefinition = '';
-
-          // If the schema is a union type (oneOf, anyOf) and includes a reference (has imports),
-          // we need to cast the property to the specific type to avoid TypeScript errors.
-          if (shouldCast && imports[0]) {
+        for (const subSchema of combinedSchemas) {
+          const { imports } = resolveSchemaRef(subSchema, context);
+          if (imports[0]) {
             additionalImports.push(imports[0]);
-            newPropName = `${propName}${pascal(imports[0].name)}`;
-            newPropDefinition = `const ${newPropName} = (${propName} as ${imports[0].name}${isRequestBodyOptional ? ' | undefined' : ''});\n`;
           }
-
-          return (
-            newPropDefinition +
-            resolveSchemaPropertiesToFormData({
+        }
+      } else {
+        const combinedSchemasFormData = combinedSchemas
+          .map((subSchema) => {
+            const { schema: combinedSchema } = resolveSchemaRef(
+              subSchema,
+              context,
+            );
+            return resolveSchemaPropertiesToFormData({
               schema: combinedSchema,
               variableName,
-              propName: newPropName,
+              propName,
               context,
               isRequestBodyOptional,
               encoding,
-            })
-          );
-        })
-        .filter(Boolean)
-        .join('\n');
+            });
+          })
+          .filter(Boolean)
+          .join('\n');
 
-      form += combinedSchemasFormData;
+        form += combinedSchemasFormData;
+      }
     }
 
     if (schema.properties) {

--- a/packages/core/src/getters/res-req-types.ts
+++ b/packages/core/src/getters/res-req-types.ts
@@ -257,6 +257,12 @@ export function getResReqTypes(
                 effectivePropName = imports[0].name;
               }
             } else if (mediaType.schema) {
+              // When schema is a oneOf/anyOf of $refs, concat schema names
+              // so the FormData variable matches the function parameter.
+              // If the union only contains inline variants (no $ref), `names`
+              // stays empty and effectivePropName keeps the default
+              // `pascal(name) + pascal(key)` form; this is the original bug's
+              // shape, but the DTO-less case is outside this fix's scope.
               const combinedRefs =
                 getSchemaOneOf(mediaType.schema) ??
                 getSchemaAnyOf(mediaType.schema);
@@ -578,10 +584,17 @@ function getSchemaFormDataAndUrlEncoded({
       if (shouldCast) {
         // If the outer schema also has direct properties, those are handled
         // below by the dedicated properties branch. Skip them here to avoid
-        // appending the same key twice.
+        // appending the same key twice. Exclude readOnly direct properties
+        // so they can still flow through the runtime loop if a variant
+        // declares the same key as writable.
         const directProperties = getSchemaProperties(schema);
         const directKeys = directProperties
-          ? Object.keys(directProperties)
+          ? Object.entries(directProperties)
+              .filter(
+                ([, value]) =>
+                  !resolveSchemaRef(value, context).schema.readOnly,
+              )
+              .map(([key]) => key)
           : [];
         const skipLine =
           directKeys.length > 0

